Add SPI to disable rich JavaScript features

szewai opened the pull request:
#### d9544bb9df1172b024bb17640a28d5b4eb0b57a4
<pre>
Add SPI to disable rich JavaScript features
<a href="https://bugs.webkit.org/show_bug.cgi?id=280908">https://bugs.webkit.org/show_bug.cgi?id=280908</a>
<a href="https://rdar.apple.com/137310813">rdar://137310813</a>

Reviewed by Ryosuke Niwa.

Declare some features as &quot;Rich JavaScript&quot; features in UnifiedWebPreferences.yaml, and add an SPI on WKPreferences
to disable them altogether.

* Source/WTF/Scripts/GeneratePreferences.rb:
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebKit/Scripts/PreferencesTemplates/WebPreferencesFeatures.cpp.erb:
* Source/WebKit/UIProcess/API/Cocoa/WKPreferences.mm:
(-[WKPreferences _disableRichJavaScriptFeatures]):
* Source/WebKit/UIProcess/API/Cocoa/WKPreferencesPrivate.h:
* Source/WebKit/UIProcess/WebPreferences.h:
* Tools/TestWebKitAPI/SourcesCocoa.txt:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKPreferences.mm: Added.
(-[WKPreferencesMessageHandler userContentController:didReceiveScriptMessage:]):
(log):
(DisableRichJavaScriptFeatures)):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKPrinting.mm:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIPermissions.mm:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WebSocket.mm:

Canonical link: <a href="https://commits.webkit.org/284787@main">https://commits.webkit.org/284787@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a927e20b23a4c1a3d3ac45a19c497a0432b3f88b

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/70498 "2 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/49904 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/23263 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/74587 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/21676 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/72615 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/57704 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/21516 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/55863 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/14334 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/73564 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/45390 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/60774 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/36325 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/70008 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/42049 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/18210 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/20037 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/63619 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/63986 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/18554 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/76306 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/69745 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/14724 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/17792 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/63585 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/14767 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/60841 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/63519 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/15613 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/11565 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/5200 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/91528 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/45706 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/473 "Built successfully") | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/19955 "Found 1 new JSC stress test failure: wasm.yaml/wasm/stress/tail-call-bbq.js.wasm-bbq (failure)") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/46780 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/48057 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/46522 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rniwa commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WKPreferences.mm:
```diff
@@ -0,0 +1,119 @@
+    RetainPtr result = (NSString *)[getNextMessage() body];
+    EXPECT_WK_STREQ(@"Loaded", result.get());
+
+    [webView evaluateJavaScript:@"canvas = document.createElement('canvas'); log(canvas.getContext('webgl') ? 'WebGL Enabled' : 'WebGL Disabled');" completionHandler:nil];
```
Why not just `log(!!canvas.getContext('webgl'))`?

szewai replied:
Because we want to know which interface fails.

rniwa replied:
oh I see.

rniwa commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WKPreferences.mm:
```diff
@@ -0,0 +1,119 @@
+
+    [webView evaluateJavaScript:@"canvas = document.createElement('canvas'); log(canvas.getContext('webgl') ? 'WebGL Enabled' : 'WebGL Disabled');" completionHandler:nil];
+    result = (NSString *)[getNextMessage() body];
+    EXPECT_WK_STREQ(@"WebGL Disabled", result.get());
```
and this could just be `EXPECT_TRUE([result boolValue])`

szewai committed:
Add SPI to disable rich JavaScript features
https://bugs.webkit.org/show_bug.cgi?id=280908
rdar://137310813

Reviewed by Ryosuke Niwa.

Declare some features as "Rich JavaScript" features in UnifiedWebPreferences.yaml, and add an SPI on WKPreferences
to disable them altogether.

* Source/WTF/Scripts/GeneratePreferences.rb:
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebKit/Scripts/PreferencesTemplates/WebPreferencesFeatures.cpp.erb:
* Source/WebKit/UIProcess/API/Cocoa/WKPreferences.mm:
(-[WKPreferences _disableRichJavaScriptFeatures]):
* Source/WebKit/UIProcess/API/Cocoa/WKPreferencesPrivate.h:
* Source/WebKit/UIProcess/WebPreferences.h:
* Tools/TestWebKitAPI/SourcesCocoa.txt:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKPreferences.mm: Added.
(-[WKPreferencesMessageHandler userContentController:didReceiveScriptMessage:]):
(log):
(DisableRichJavaScriptFeatures)):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKPrinting.mm:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIPermissions.mm:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WebSocket.mm:

Canonical link: https://commits.webkit.org/284787@main
